### ISSUE TITLE
Release process for WasmEdge 0.14.0

### ISSUE THREAD

**Issue by q82419** [WasmEdge/WasmEdge#3005]
### Summary

**Planned alpha.1 date: 2024/01/05**

**Planned alpha.2 date: 2024/02/23**

**Planned alpha.3 date: 2024/03/05**

**Planned rc.1 date: 2024/03/15**

**Planned rc.2 date: 2024/03/20**

**Planned rc.3 date: 2024/03/22**

**Planned rc.4 date: 2024/04/03**

**Planned alpha.4 date: 2024/04/30**

**Planned rc.5 date: 2024/05/08**

**Planned release date: TBD**

### Features

- [x] WASM serialization
   - #2819
- [x] WebAssembly Typed Function Reference Proposal
   - #2691
   - #3129
- [x] Loader phase of WebAssembly Component Model Proposal
   - #3036
   - #3149 
   - #3152
- [x] Preparation of WebAssembly GC Proposal
   - #3094 
- [x] WebAssembly GC Proposal
   - #3007
   - #3072
- [x] WebAssembly Exception Handling Proposal
   - #3306 

### Issues

- [ ] RustLS plugin fail due to the ABI change
   - #3138

### Documents

- [x] Update the C API document.

**Comment by bashor** [comment]
@hydai @q82419 WasmEdge `0.14.0-alpha.2` doesn't work on macOS, at least due to a lack of `libprotobuf.25.1.0` and `libgrpc++.1.60.dylib`.

**Comment by hydai** [comment]
Hi @bashor
Thanks for raising this issue; I am working on fix #3242 and will have `alpha.3` soon to fix it.

**Comment by bashor** [comment]
@hydai @q82419 Any ETA for alpha.3? Do you have any plans for EH proposal in 0.14.0?

**Comment by hydai** [comment]
We have Peace Memorial Day in Taiwan this week, so the `alpha.3` will be released next week.
The EH proposal is not in 0.14.0; we are going to have 0.14.1 once the EH proposal finishes the implementation.

**Comment by bashor** [comment]
Hi, @hydai!
Good news, I managed to run a sample compiled from Kotlin in WasmEdge `0.14.0-rc.1` and print something to the stdout.
But it required some tricks to avoid initialization, since otherwise WasmEdge fails with:

```
segmentation fault  wasmedge --enable-all --reactor  one
```

**Comment by hydai** [comment]
Hi @bashor 
May I have the WASM to debug this issue?

**Comment by bashor** [comment]
Here it is [kotlin-wasm-wasi-example-wasm-wasi.wasm.zip](https://github.com/WasmEdge/WasmEdge/files/14635149/kotlin-wasm-wasi-example-wasm-wasi.wasm.zip)

Following works by running `_start`, which directly calls wasi APIs
```shell
$ wasmedge --enable-all kotlin-wasm-wasi-example-wasm-wasi.wasm    
[2024-03-18 12:02:01.300] [warning] GC proposal is enabled, this is experimental.
[2024-03-18 12:02:01.301] [warning] component model is enabled, this is experimental.
Hi!%  
```

This fails at startup during initialization, in general, this initialization part is required for Kotlin/Wasm programs:
```shell
$ wasmedge --enable-all --reactor kotlin-wasm-wasi-example-wasm-wasi.wasm one
[2024-03-18 12:03:21.796] [warning] GC proposal is enabled, this is experimental.
[2024-03-18 12:03:21.797] [warning] component model is enabled, this is experimental.
zsh: segmentation fault  wasmedge --enable-all --reactor  one
```

**Comment by bashor** [comment]
@q82419 Thanks for the fix! When do you expect it to be available?

**Comment by hydai** [comment]
@q82419 
Let's have rc.2 tmr once #3292 got merged.

@bashor, Would you have more examples to test the GC proposal? We found that the above error is not included in the GC spec tests. We would like to have more examples and tests to verify the GC proposal before we have the official 0.14.0. WDYT?

**Comment by bashor** [comment]
@hydai Right now, we have only a basic example [here](https://github.com/Kotlin/kotlin-wasm-examples/tree/main/wasi-example).

**Comment by bashor** [comment]
As soon as we get a build with #3292 we can try to run our compiler tests, but it might be tricky since so far we run tests only on JS + Wasm VMs.

**Comment by q82419** [comment]
Hi @bashor ,

[rc.2](https://github.com/WasmEdge/WasmEdge/releases/tag/0.14.0-rc.2) released and welcome to test with it.
And also feel free to report if there's bug in GC implementation.
Thanks!

**Comment by bashor** [comment]
@q82419 Thanks! 
I've checked rc.2 with simple examples, and it works well so far.

**Comment by bashor** [comment]
Recently I ran a big chunk of compiler tests (7561) with rc.2, and about 90% of them passed. 

I see two big groups of tests that failed: 
* Tests in the first group failed on unreachable instruction which was generated instead of throw instruction due to the lack of exception handling support.
* Tests in the second group failed due to a problem in our test infra.

Good job! 🙌 
Congrats, folks! 🎉

cc @q82419 @hydai @juntao

**Comment by hydai** [comment]
Thanks @bashor 
That's such a good news. We are working on the Exception Handling proposal now. We would like to have it in the 0.14.1 soon.

**Comment by bashor** [comment]
> We are working on the Exception Handling proposal now. We would like to have it in the 0.14.1 soon.
Great! Do you have an approximate timeline for this?

**Comment by hydai** [comment]
A working PR is here: https://github.com/WasmEdge/WasmEdge/pull/3306
We would like to have it at the end of April or later.

**Comment by q82419** [comment]
Hi @bashor ,

We planned to back to alpha.4 in these days for adding the exception-handling proposal supporting in `0.14.0` after merging #3306 .
Please feel free to test then. Thanks!

**Comment by bashor** [comment]
@q82419 🔥 Super! Looking forward to it!

**Comment by bashor** [comment]
@q82419 As I see, you are supporting only the latest version of EH, and I'm wondering if you could support a previous version as well, maybe a limited version w/o rethrow. 

Reasoning:
* The previous version of the proposal has been available by default in all browsers for quite a long time now
* In Kotlin, we still use the previous version of the proposal by default and can only switch to the new one once all major VMs support it and most users migrate to those VM versions.
  * Likely, it is the same for other languages using the proposal.

**Comment by q82419** [comment]
> @q82419 As I see, you are supporting only the latest version of EH, and I'm wondering if you could support a previous version as well, maybe a limited version w/o rethrow.
> 
> Reasoning:
> 
> * The previous version of the proposal has been available by default in all browsers for quite a long time now
> * In Kotlin, we still use the previous version of the proposal by default and can only switch to the new one once all major VMs support it and most users migrate to those VM versions.
>   
>   * Likely, it is the same for other languages using the proposal.

Sorry that we just updated the latest version (https://github.com/WebAssembly/exception-handling/pull/285) and tuned the performance.
The old one still remain on the commit (https://github.com/WasmEdge/WasmEdge/pull/3306/commits/74ad572a72b23c1925731b7cfaa940553f038389), but the performance is not tuned, and still some spec test failed.
It will take times if we support the both versions.

cc @hydai

**Comment by bashor** [comment]
@q82419, any updates on EH and alpha.4?

**Comment by q82419** [comment]
Hi @bashor ,

I'm working with supporting the legacy exception-handling proposal, which is supported by the toolchain.
I think it takes days. After passing the spec tests, we'll have the alpha.4 with the previous version of EH.
Thanks!

**Comment by bashor** [comment]
🤩 Wow! Nice! Thanks for the efforts!

**Comment by bashor** [comment]
I'm just wondering if `alpha.4` will have support for both versions of EH.

**Comment by bashor** [comment]
Just in case, `rethrow` is not needed for Kotlin (I've heard it's tricky to support in old EH)

**Comment by q82419** [comment]
> I'm just wondering if `alpha.4` will have support for both versions of EH.

Sure, the both version of EH will be supported.
We'll remove the legacy one once the proposal entering phase 4, or the toolchain updated.

**Comment by q82419** [comment]
> Just in case, `rethrow` is not needed for Kotlin (I've heard it's tricky to support in old EH)

In current status, the EH instructions are done except the `delegate` and `rethrow`.
Thanks for the information, and we'll consider this.

**Comment by bashor** [comment]
JFTR: we don't use `delegate` too :)

**Comment by hydai** [comment]
@q82419 
Could we ignore `delegate` and `rethrow`? The only reason that we want to support the old EH is for Kotlin. If Kotlin doesn't need them, leaving these features alone is okay. The old one will deprecate once the toolchain is updated to the new EH.

**Comment by q82419** [comment]
Sure. I'll remove the related tests.

**Comment by hydai** [comment]
Let's have alpha.4 once the PR is ready. From now on, ignore the `delegate` and `rethrow`.

**Comment by q82419** [comment]
Hi @bashor ,
We've updated the `alpha.4` pre-release.
Please feel free to test with it and report if something wrong.
Thanks!

**Comment by bashor** [comment]
@q82419 thanks! We'll check it out.

**Comment by bashor** [comment]
Great job, folks! I ran ~7500 tests, and only two issues were found (4 tests failed).

**Comment by bashor** [comment]
Unfortunately, I didn't have much time to minimize issues, so I share what I have so far.

Examples were run with the following command:
```
 'wasmedge' '--enable-gc' '--enable-exception-handling' 'dev/index.wasm' 'runBoxTest'
```

[exceptionInFinally.zip](https://github.com/WasmEdge/WasmEdge/files/15170814/exceptionInFinally.zip)
Output:
```
[2024-04-30 22:33:20.192] [warning] GC proposal is enabled, this is experimental.
Uncaught exception: AssertionError: failed on 1. Expected <called->exception->finally->closed>, actual <called->exception->finally->closedfinally->closed>.
[2024-04-30 22:33:20.199] [error] execution failed: uncaught exception, Code: 0x419
[2024-04-30 22:33:20.199] [error]     When executing function name: "runBoxTest"
```

[tryCatchWithHandleResult.zip](https://github.com/WasmEdge/WasmEdge/files/15170817/tryCatchWithHandleResult.zip)
Output:
```
[2024-04-30 22:44:44.250] [warning] GC proposal is enabled, this is experimental.
Uncaught exception: NullPointerException
[2024-04-30 22:44:44.263] [error] execution failed: uncaught exception, Code: 0x419
[2024-04-30 22:44:44.263] [error]     When executing function name: "runBoxTest"
```

Normally, an output should be something like this:
```
[2024-04-30 22:45:44.738] [warning] GC proposal is enabled, this is experimental.
1
```

Also, in the examples, you can find `test.mjs` files which could be used to run tests in the latest Node.js.

**Comment by q82419** [comment]
Hi @bashor ,

Thanks for your test WASM, and I can fix a bug for EH proposal.
But after fixing that, it still output the same error: `AssertionError` for the `exceptionInFinally`, and `NullPointerException` for the `tryCatchWithHandleResult`.

The test WASM are too large that it's hard to trace through the binary.
Would you help to provide the source code of these WASM files?
Thanks!

**Comment by q82419** [comment]
> @q82419 Originally files from our compiler tests:
> 
> * (exceptionInFinally.kt)[https://github.com/JetBrains/kotlin/blob/z/wasmedge/compiler/testData/codegen/boxInline/nonLocalReturns/tryFinally/exceptionTable/exceptionInFinally.kt]
> * (tryCatchWithHandleResult.kt)[https://github.com/JetBrains/kotlin/blob/z/wasmedge/compiler/testData/codegen/box/coroutines/tryCatchWithHandleResult.kt]
> 
> I think it would be easier for you to work with a small project so I moved relevant sources to a separate project and modified a bit. Check out [wasi-example](https://github.com/Kotlin/kotlin-wasm-examples/tree/wasmedge/wasi-example) folder in `wasmedge` branch of https://github.com/Kotlin/kotlin-wasm-examples
> 
> You can build the project and run a wasm file in wasmedge with the following command:
> 
> ```
> ./gradlew compileDevelopmentExecutableKotlinWasmWasi && wasmedge --enable-gc --enable-exception-handling build/compileSync/wasmWasi/main/developmentExecutable/kotlin/kotlin-wasm-wasi-example-wasm-wasi.wasm runTests
> ```
> 
> Normally, it should print:
> 
> ```
> Start...
> 
> # exceptionInFinally
> OK
> 
> # tryCatchWithHandleResult
> OK
> 
> Done
> ```
> 
> But with WasmEdge, it prints:
> 
> ```
> Start...
> 
> # exceptionInFinally
> Wrong result:[called->exception->finally->closedfinally->closed], expected: [called->exception->finally->closed]
> 
> # tryCatchWithHandleResult
> ERROR: NullPointerException
> 
> Done
> ```

Hi @bashor ,

Thanks for the information.
I found the bug and can run these tests successfully.
But for fixing this, the performance tuning is neccessary.
We'll fix that and publish a `rc` version in these days.
Thanks!

**Comment by bashor** [comment]
@q82419 great to hear that, thanks for the update!

**Comment by q82419** [comment]
This will be fixed in #3387 .

**Comment by q82419** [comment]
Hi @bashor ,

`rc.5` is published. Please feel free to test with it and report if there's something wrong.
Thanks!

**Comment by q82419** [comment]
@L-jasmine , @apepkuss ,

Please help to update the rust SDK base on the `0.14.0-rc.5`. There's no API change after this rc version. Thanks!

**Comment by bashor** [comment]
@q82419 Thanks for the update! So far, no issues with `rc.5`, it passed all compiler tests I tried before (~7500).

**Comment by hydai** [comment]
> @q82419 Thanks for the update! So far, no issues with `rc.5`, it passed all compiler tests I tried before (~7500).

Awesome! Good to know that. Thanks for the information.
@q82419 Let's prepare the official release next week.
